packages/@aws-cdk/aws-cloudformation/lib/custom-resource.ts

```diff
@@ -126,3 +107,24 @@ function uppercaseProperties(props: Properties): Properties {
   });
   return ret;
 }
+
+function renderResourceType(resourceType?: string) {
+  if (!resourceType) {
+    return CfnCustomResource.resourceTypeName;
+  }
+
+  if (!resourceType.startsWith('Custom::')) {
+    throw new Error(`Custom resource type must begin with "Custom::" (${resourceType})`);
+  }
+
+  const typeName = resourceType.substr(resourceType.indexOf('::') + 2);
+  if (typeName.length > 60) {
+    throw new Error(`Custom resource type length > 60 (${resourceType})`);
+  }
+
+  if (!/^[a-z0-9_@-]+$/i.test(typeName)) {
+    throw new Error(`Custom resource type name can only include alphanumeric characters and _@- (${typeName})`);
+  }
```

Comment:
Could have checked this whole thing in a single pass:

```
resourceType.match(/^Custom::[a-z0-9_@-]{1,60}$/);
```

Reply:
This was moved, I prefer not to dig into it right now.